### ISSUE TITLE
assigning a class with __call__ to dunder method shows the wrong type

### ISSUE THREAD

**Issue by adhami3310** [microsoft/pyright#9470]
```python
from typing_extensions import Self, reveal_type


class Function:
    def __call__(self, *args, **kwargs) -> Self:
        return self


class ArrayLike:
    __getitem__ = Function()


x = ArrayLike()[0]
reveal_type(x) # Type of "x" is "ArrayLike"
print(type(x))
```
but prints: `<class '__main__.Function'>`, as self is `Function`. This doesn't feel like intended behavior.

I'm using `pyright 1.1.389`.

**Comment by adhami3310** [comment]
Interestingly, `x = ArrayLike().__getitem__(0)` does show `x: Function`. So there's a discrepancy, even though the two are equivalent.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.390.
